File: army/src/main/java/com/rolan/army/ArmyBuilder.java
package com.rolan.army;

import android.content.Context;

import androidx.annotation.NonNull;

import java.util.Collections;
import java.util.List;

/**
 * Created by Rolan.Wang on 1/28/21.5:49 PM
 * 描述：
 */
public final class ArmyBuilder {

    private List<RequestListener> defaultRequestListeners;
    private Engine engine;

    Army builder(@NonNull Context context){
        if (defaultRequestListeners == null) {
            defaultRequestListeners = Collections.emptyList();
        } else {
            defaultRequestListeners = Collections.unmodifiableList(defaultRequestListeners);
        }

        RequestManagerRetriever requestManagerRetriever =
                new RequestManagerRetriever();
        if (engine == null) {
            engine = new Engine();
        }
        return new Army(context,defaultRequestListeners,requestManagerRetriever,engine);
    }
}

File: army/src/main/java/com/rolan/army/ArmyBuilder.kt
package com.rolan.army

import android.content.Context
import com.rolan.army.RequestListener
import com.rolan.army.IEngine
import com.rolan.army.ArmyBuilder
import com.rolan.army.Army
import com.rolan.army.RequestManagerRetriever
import com.rolan.army.DefaultEngine
import java.util.*
import kotlin.collections.ArrayList

/**
 * Created by Rolan.Wang on 1/28/21.5:49 PM
 * Army instance construction, externally recreable
 */
class ArmyBuilder {
    private var defaultRequestListeners: MutableList<RequestListener>? = null
    private var engine: IEngine? = null

    /**
     * Add a global listener. Each request will be carried with this callback
     * @param listener
     * @return
     */
    fun addGlobalRequestListener(listener: RequestListener): ArmyBuilder {
        if (defaultRequestListeners == null) {
            defaultRequestListeners = ArrayList()
        }
        defaultRequestListeners?.add(listener)
        return this
    }

    /**
     * Set up the custom engine
     * @param engine custom engine
     * @return
     */
    fun setEngine(engine: IEngine?): ArmyBuilder {
        this.engine = engine
        return this
    }

    fun build(context: Context): Army {
        defaultRequestListeners = if (defaultRequestListeners == null) {
            ArrayList()
        } else {
            Collections.unmodifiableList(defaultRequestListeners)
        }
        val requestManagerRetriever = RequestManagerRetriever()
        if (engine == null) {
            engine = DefaultEngine()
        }
        return Army(context, defaultRequestListeners!!, requestManagerRetriever, engine!!)
    }
}